core/src/main/java/org/infinispan/stream/impl/DistributedCacheStream.java

```diff
@@ -537,23 +470,130 @@ public long count() {
+         }
+      });
+   }
+
+   private void rehashAwareIteration(AtomicBoolean complete, Consumer<R> consumer, IteratorSupplier<R> supplier, boolean iteratorParallelDistribute) {
+      ConsistentHash segmentInfoCH = dm.getReadConsistentHash();
+      SegmentListenerNotifier<R> listenerNotifier;
+      if (segmentCompletionListener != null) {
+         listenerNotifier = new SegmentListenerNotifier<>(
+                 segmentCompletionListener);
+         supplier.setConsumer(listenerNotifier);
+      } else {
+         listenerNotifier = null;
       }
+      KeyTrackingConsumer<Object, R> results = new KeyTrackingConsumer<>(segmentInfoCH,
+              iteratorOperation.wrapConsumer(consumer), iteratorOperation.getFunction(),
+              listenerNotifier);
+      Thread thread = Thread.currentThread();
+      executor.execute(() -> {
+         try {
+            log.tracef("Thread %s submitted iterator request for stream", thread);
+            Set<Integer> segmentsToProcess = segmentsToFilter == null ?
+                    new ReplicatedConsistentHash.RangeSet(segmentInfoCH.getNumSegments()) : segmentsToFilter;
+            do {
+               ConsistentHash ch = dm.getReadConsistentHash();
+               boolean runLocal = ch.getMembers().contains(localAddress);
+               Set<Integer> segments;
+               Set<Object> excludedKeys;
+               boolean stayLocal = false;
+               if (runLocal) {
+                  Set<Integer> segmentsForOwner = ch.getSegmentsForOwner(localAddress);
+                  stayLocal = segmentsToFilter != null && segmentsForOwner.containsAll(segmentsToFilter);
+                  segments = stayLocal ? segmentsForOwner : ch.getPrimarySegmentsForOwner(localAddress);
+                  segments.retainAll(segmentsToProcess);
+
+                  excludedKeys = segments.stream().flatMap(s -> results.referenceArray.get(s).stream())
+                          .collect(Collectors.toSet());
+               } else {
+                  segments = null;
+                  excludedKeys = Collections.emptySet();
+               }
+               KeyTrackingTerminalOperation<Object, R, Object> op = iteratorOperation.getOperation(
+                       intermediateOperations, supplierForSegments(ch, segmentsToProcess, excludedKeys, !stayLocal),
+                       distributedBatchSize);
+               if (!stayLocal) {
+                  UUID id = csm.remoteStreamOperationRehashAware(iteratorParallelDistribute, parallel, ch,
+                          segmentsToProcess, keysToFilter, new AtomicReferenceArrayToMap<>(results.referenceArray),
+                          includeLoader, op, results);
+                  supplier.pending = id;
+                  try {
+                     if (runLocal) {
+                        performLocalRehashAwareOperation(results, segmentsToProcess, ch, segments, op,
+                                () -> ch.getPrimarySegmentsForOwner(localAddress), id);
```

Comment:
Same here with getPrimarySegmentsForOwner, you can just pass segments as is since the way the logic works it will always be the primary owners.